ykrt/src/compile/jitc_yk/codegen/x64/mod.rs

```diff
@@ -1424,97 +1438,105 @@ impl<'a> Assemble<'a> {
-                let [_] = self.ra.assign_fp_regs(
-                    &mut self.asm,
-                    iidx,
-                    [RegConstraint::OutputFromReg(Rx::XMM0)],
-                );
+                let cnstr = match &fp_cnstrs[0] {
+                    RegConstraint::InputIntoRegAndClobber(op, _) => {
+                        RegConstraint::InputOutputIntoReg(op.clone(), Rx::XMM0)
+                    }
+                    RegConstraint::Clobber(_) => RegConstraint::OutputFromReg(Rx::XMM0),
+                    _ => unreachable!(),
+                };
+                fp_cnstrs[0] = cnstr;
             }
             Ty::Integer(_) | Ty::Ptr => {
-                let [_] = self.ra.assign_gp_regs(
-                    &mut self.asm,
-                    iidx,
-                    [RegConstraint::OutputFromReg(Rq::RAX)],
-                );
+                let rax_i = CALLER_CLOBBER_REGS
+                    .iter()
+                    .position(|x| *x == Rq::RAX)
+                    .unwrap();
+                gp_cnstrs[rax_i] = RegConstraint::OutputFromReg(Rq::RAX)
             }
             Ty::Func(_) => todo!(),
             Ty::Unimplemented(_) => todo!(),
         }
 
-        if fty.is_vararg() {
-            // SysV x64 ABI says "rax is used to indicate the number of vector arguments passed
-            // to a function requiring a variable number of arguments". Float arguments are passed
-            // in vector registers.
-            dynasm!(self.asm; mov rax, num_float_args);
-        }
+        // We now have all the FP constraints, so assign those.
+        let _: [Rx; 16] =
+            self.ra
+                .assign_fp_regs(&mut self.asm, iidx, fp_cnstrs.try_into().unwrap());
+        let num_float_args = i32::try_from(num_float_args).unwrap();
 
-        // Actually perform the call.
+        // We now have most of the GP constraints, except the call target. We have to handle that
+        // differently, depending on whether this is a direct or indirect call.
         match (callee, callee_op) {
             (Some(p), None) => {
-                let [reg] = self.ra.assign_gp_regs_avoiding(
-                    &mut self.asm,
-                    iidx,
-                    [RegConstraint::Temporary],
-                    RegSet::from_vec(&CALLER_CLOBBER_REGS),
-                );
+                // Direct call
+                gp_cnstrs.push(RegConstraint::Temporary);
+                let [_, _, _, _, _, _, _, _, _, tmp_reg] =
+                    self.ra
+                        .assign_gp_regs(&mut self.asm, iidx, gp_cnstrs.try_into().unwrap());
+
+                if fty.is_vararg() {
+                    dynasm!(self.asm; mov rax, num_float_args); // SysV x64 ABI
+                }
                 dynasm!(self.asm
-                    ; mov Rq(reg.code()), QWORD p as i64
-                    ; call Rq(reg.code())
+                    ; mov Rq(tmp_reg.code()), QWORD p as i64
```

Comment:
Side note: it's been on my todo list to optimise the way we do the actual call.

It should be possible to use one of the encodings which doesn't require using a register, but last time I tried (ages ago) it wasn't as simple as I thought.

Reply:
Agreed: I want this too. Also, when we can't do so, we often (but not always!) have a spare register such that we don't have to clobber another random register.